meilisearch-lib/src/index/search.rs

```diff
@@ -436,50 +445,71 @@ fn make_document(
     field_ids_map: &FieldsIdsMap,
     obkv: obkv::KvReaderU16,
 ) -> Result<Document> {
-    let mut document = Document::new();
+    let mut document = serde_json::Map::new();
 
-    for attr in attributes_to_retrieve {
-        if let Some(value) = obkv.get(*attr) {
-            let value = serde_json::from_slice(value)?;
+    // recreate the original json
+    for (key, value) in obkv.iter() {
+        let value = serde_json::from_slice(value)?;
+        let key = field_ids_map
+            .name(key)
+            .expect("Missing field name")
+            .to_string();
 
-            // This unwrap must be safe since we got the ids from the fields_ids_map just
-            // before.
-            let key = field_ids_map
-                .name(*attr)
-                .expect("Missing field name")
-                .to_string();
-
-            document.insert(key, value);
-        }
+        document.insert(key, value);
     }
+
+    // select the attributes to retrieve
+    let attributes_to_retrieve = attributes_to_retrieve
+        .iter()
+        .map(|&fid| field_ids_map.name(fid).expect("Missing field name"));
+
+    let document = permissive_json_pointer::select_values(&document.into(), attributes_to_retrieve);
+
+    // then we need to convert the `serde_json::Map` into an `IndexMap`.
+    let document = document.into_iter().collect();
```

Comment:
Yes, I was concerned about that too. From what I see it works but it's not specified it should so I guess I'm just lucky.
I would like to move the `permissive-json-pointer` crate into meilisearch so we could make it work directly with the `IndexMap`.
Do you think this needs to be fixed before the RC0 or can it wait until the RC1?

@gmourier @curquiza 
Basically, the jsons in `hits` could have their fields in a different order each time.

Reply:
actually, when the `preserve_order` feature from serde_json is enables, a `serde_json::Map` is nothing but an `IndexMap` under the hood: https://docs.serde.rs/src/serde_json/map.rs.html#23-26

and it turns out that we're lucky: https://github.com/meilisearch/meilisearch/blob/main/meilisearch-lib/Cargo.toml#L45

This now begs the question: Do we need to do the back and forth conversion?

Reply:
Ah! No we don't need and since that's the case I guess we should just remove entirely the `IndexMap` and define a `Document` as a `serde_json::Map<String, Value>`!

Reply:
I think we can indeed 🤔

Reply:
> Basically, the jsons in hits could have their fields in a different order each time.

Ok to wait for rc1 to fix this!